m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,765 @@
+    @Override
+    public void reportGauge(String name, Map<String, String> tags, double value) {
+        GaugeValue gaugeValue = new GaugeValue();
+        gaugeValue.setDValue(value);
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setGauge(gaugeValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    @Override
+    public void reportTimer(String name, Map<String, String> tags, Duration interval) {
+        TimerValue timerValue = new TimerValue();
+        timerValue.setI64Value(interval.getNanos());
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setTimer(timerValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    @Override
+    public void reportHistogramValueSamples(
+        String name,
+        Map<String, String> tags,
+        Buckets buckets,
+        double bucketLowerBound,
+        double bucketUpperBound,
+        long samples
+    ) {
+        // Append histogram bucket-specific tags
+        int bucketIdLen = String.valueOf(buckets.size()).length();
+        bucketIdLen = Math.max(bucketIdLen, MIN_METRIC_BUCKET_ID_TAG_LENGTH);
+
+        String bucketIdFmt = String.format("%%0%sd", bucketIdLen);
+
+        BucketPair[] bucketPairs = BucketPairImpl.bucketPairs(buckets);
+
+        if (tags == null) {
+            tags = new HashMap<>();
+        }
+
+        for (int i = 0; i < bucketPairs.length; i++) {
+            // Look for the first pair with an upper bound greater than or equal
+            // to the given upper bound.
+            if (bucketPairs[i].upperBoundValue() >= bucketUpperBound) {
+                String idTagValue = String.format(bucketIdFmt, i);
+
+                tags.put(bucketIdTagName, idTagValue);
+                tags.put(bucketTagName,
+                    String.format("%s-%s",
+                        valueBucketString(bucketPairs[i].lowerBoundValue()),
+                        valueBucketString(bucketPairs[i].upperBoundValue())
+                    )
+                );
+
+                break;
+            }
+        }
+
+        reportCounterInternal(name, tags, samples);
+    }
+
+    @Override
+    public void reportHistogramDurationSamples(
+        String name,
+        Map<String, String> tags,
+        Buckets buckets,
+        Duration bucketLowerBound,
+        Duration bucketUpperBound,
+        long samples
+    ) {
+        // Append histogram bucket-specific tags
+        int bucketIdLen = String.valueOf(buckets.size()).length();
+        bucketIdLen = Math.max(bucketIdLen, MIN_METRIC_BUCKET_ID_TAG_LENGTH);
+
+        String bucketIdFmt = String.format("%%0%sd", bucketIdLen);
+
+        BucketPair[] bucketPairs = BucketPairImpl.bucketPairs(buckets);
+
+        if (tags == null) {
+            tags = new HashMap<>(2, 1);
```

Comment:
use const vars rather than literals heres

Reply:
These numbers are used for their specific circumstances. If `tags == null`, then we know that there will only be two items in the map (the two that we add later on). The 2 is the initial capacity of the buckets, and the 1 defines when to rehash.